### PR TITLE
[CDAP-18785] Add metrics to flow-control

### PR DESCRIPTION
**Pull request by masoud-io**
Adding two metrics to flow-control:
1. gauge metric for launching programs
2. gauge metric for running programs

Also modifies the logic for counting running programs to only count workflows program type (instead of all program types).

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 1]
```diff
@@ -152,6 +151,15 @@ public void removeRequest(ProgramRunId programRunId) {
...
                programRunId, launchingQueue.size());
     }
+
+    int launchingCount = launchingQueue.size();
+    emitMetrics(Constants.Metrics.FlowControl.FLOW_CONTROL_LAUNCHING_RUNS, launchingCount);
+    emitMetrics(Constants.Metrics.FlowControl.FLOW_CONTROL_OUTSTANDING_RUNS, launchingCount + getRunningCount());
+  }
+
+  private void emitMetrics(String metricName, long value) {
+    Map<String, String> metricTags = new HashMap<>();
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: Why not simply `Collections.emptyMap()`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 2]
```diff
@@ -166,11 +174,23 @@ private void cleanupQueue() {
       // Queue head might have already been removed. So instead of calling poll, we call remove.
       if (launchingQueue.remove(programRunId)) {
         LOG.info("Removing request with runId {} due to expired retention time.", programRunId);
+        emitMetrics(Constants.Metrics.FlowControl.FLOW_CONTROL_LAUNCHING_RUNS, launchingQueue.size());
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: I think outstanding will also go down

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 3]
```diff
@@ -166,11 +174,23 @@ private void cleanupQueue() {
...
       }
-
     }
   }
 
+  private int getRunningCount() {
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: Is it cheap enough? We now call it each time we add / remove and item from launching runs. If it's not cheap, it's better to have "running" metric that we don't need to update when run is added /removed from launching

**Review by tivv — COMMENTED** [review]
Where do we reduce FLOW_CONTROL_OUTSTANDING_RUNS when run finishes or program starts? Also, why do we limit programs at all (and not only workflows)? Each workflow has at least one MAPREDUCE/SPARK programs, so we are effectively reducing limit twice this way

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java (hunk 2 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: Thanks for catching this. Decided to change outstanding to only running metrics for simplicity.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java (hunk 3 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: Thanks. running metrics no longer is emitted when a launch queue changes.

**Comment by masoud-io** [comment]
> Where do we reduce FLOW_CONTROL_OUTSTANDING_RUNS when run finishes or program starts? Also, why do we limit programs at all (and not only workflows)? Each workflow has at least one MAPREDUCE/SPARK programs, so we are effectively reducing limit twice this way

Right now we have two metrics:
1. launching metric
2. running metric

therefore (1) + (2) equals total number of outstanding launch requests in the system. 

Also changed the code to only count running workflows (instead of all programs).

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java [hunk 4]
```diff
@@ -926,6 +926,14 @@ public static String getZKQuorum(CConfiguration cConf) {
       public static final long MAX_MINUTE_RESOLUTION_QUERY_INTERVAL = 600;
     }
 
+    /**
+     * Flow control metrics
+     */
+    public static final class FlowControl {
+      public static final String WORKFLOWS_LAUNCHING = "flow.control.workflows.launching";
```
Comment (greeshmaswaminathan) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java]: nit: suffix ".count" in the name to make the metric more obvious.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 5]
```diff
@@ -115,20 +120,12 @@ public Count addRequestAndGetCount(ProgramRunId programRunId) throws Exception {
...
-
-    int runsCount = (int)
-      list.stream().filter(r -> isRunning(r.getController().getState().getRunStatus())).count()
-      + launchingCount;
-
-    LOG.info("Counter has {} concurrent launching and {} concurrent runs.", launchingCount, runsCount);
-    return new Count(launchingCount, runsCount);
+    int runningCount = getWorkflowsRunningCount();
+
+    emitMetrics(Constants.Metrics.FlowControl.WORKFLOWS_LAUNCHING, launchingCount);
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: This metric is already added in addRequest() . Did you mean to add WORKFLOWS_RUNNING?

Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: Yes, it does not look like additional emit is needed at all.

**Review by tivv — APPROVED** [review]
Other than extra emit, lgtm. Please also address Greeshma's comments

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java (hunk 4 above)
Comment (masoud-io) [reply cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java]: Great idea. Done!

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java (hunk 5 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: Thanks for the catch. Done!

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 6]
```diff
@@ -138,22 +135,33 @@ public Count addRequestAndGetCount(ProgramRunId programRunId) throws Exception {
    */
   public void addRequest(ProgramRunId programRunId) {
     launchingQueue.add(programRunId);
+    emitMetrics(Constants.Metrics.FlowControl.WORKFLOWS_LAUNCHING, launchingQueue.size());
     LOG.info("Added request with runId {}.", programRunId);
   }
 
   /**
    * Remove the request with the provided programRunId when the request is no longer lunching.
```
Comment (seanzhougoogle) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: nit: lunching -> launching.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 7]
```diff
@@ -138,22 +135,33 @@ public Count addRequestAndGetCount(ProgramRunId programRunId) throws Exception {
...
    * I.e., not in-flight, not {@link ProgramRunStatus#PENDING} and not {@link ProgramRunStatus#STARTING}
    *
-   * @param programRunId
+   * @param programRunId      of the request to be removed from launching queue.
+   * @param emitRunningChange if true, also updates {@link @Constants.Metrics.FlowControl.WORKFLOWS_RUNNING}
```
Comment (seanzhougoogle) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: nit: extra `@` before `Constants` ?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 8]
```diff
@@ -138,22 +135,33 @@ public Count addRequestAndGetCount(ProgramRunId programRunId) throws Exception {
    */
   public void addRequest(ProgramRunId programRunId) {
```
Comment (seanzhougoogle) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java:134]: nit: would it be better to name it `addLaunchingRequest` ? similar comments to `addRequestAndGetCount` and `removeRequest`

**Review by seanzhougoogle — APPROVED** [review]
LGTM. Thanks Masoud :)

One question, we will have a separate PR for `rejected` metrics ?

**Comment by masoud-io** [comment]
> LGTM. Thanks Masoud :)
> 
> One question, we will have a separate PR for `rejected` metrics ?

Rejected metrics are already supported in CDAP: `system.program.rejected.runs`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java (hunk 8 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java:134]: This is a great idea. The main reason I didn't do it is `addLaunchingRequestAndGetCount` becomes quite long.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java [hunk 9]
```diff
@@ -166,9 +172,26 @@ private void cleanupQueue() {
...
       }
+    }
+  }
+
+  private int getWorkflowsRunningCount() {
+    List<ProgramRuntimeService.RuntimeInfo> list = runtimeService
+      .listAll(ProgramType.WORKFLOW);
 
+    try {
+      throw new Exception("dummy exception");
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordMonitorService.java]: I guess this has come in accidentally?

**Review by greeshmaswaminathan — APPROVED** [review]

**Pull request merged**